x-pack/plugins/security/public/authentication/login/login_page.tsx

```diff
@@ -183,6 +204,55 @@ export class LoginPage extends Component<Props, State> {
       );
     }
 
+    if (!isCookiesEnabled) {
+      if (isWindowEmbedded()) {
+        return (
+          <EuiEmptyPrompt
+            body={
+              <p>
+                <FormattedMessage
+                  id="xpack.security.loginPage.requiresNewWindowTitle"
+                  defaultMessage="This content can't be viewed embedded in another website."
```

Comment:
This was purely a UX consideration since it's recoverable and I didn't want to phrase it as an error. The technical details here matter to us but not to users who just want to get on and see the content. 

The scenario were Kibana isn't embedded and still unable to set cookies is not recoverable so we need to explain to the user what exactly went wrong and how they can fix it. I don't know how common it is for users to have cookies disabled but I imagine it to be very unlikely to happen unless we count crawlers.

Reply:
> This was purely a UX consideration since it's recoverable and I didn't want to phrase it as an error. The technical details here matter to us but not to users who just want to get on and see the content.

That's a fair point, but I'm not sure if these details don't matter to the user since the issue might be on their end (`Prevent cross-site tracking` in Safari, or `Strict` browser privacy mode in Firefox). Having this escape hatch is great, but it's still a sub-optimal UX (especially if you have multiple Kibana dashboards embedded on the same page). In this case we'd want users to understand what's going on and fix it if possible. 

Cookie and browser privacy controls aren't necessarily super technical these days, all browser vendors try to publish helpful docs on how one can tune them.

I'm curious if we had similar cases elsewhere @elastic/kibana-docs? Do we prefer to not overload users with "how-they-can-potentially-fix-the-problem-themselves" information in our error messages?

Reply:
The issue is that those users viewing the dashboard can't fix this without disabling "Prevent cross-site tracking" though and I don't think that's something we should recommend they do. 

The best way to prevent this from happening is to warn users at the point of embedding a visualisation, which requires authentication, that this will cause issues with privacy settings in modern browsers (Chrome will add this feature too) and that they should only embed visualisation on external websites using anonymous access.

Reply:
> The issue is that those users viewing the dashboard can't fix this without disabling "Prevent cross-site tracking" though and I don't think that's something we should recommend they do.

Yeah, I didn't mean we should recommend users to disable the protection mode (they can add a single exception just for the particular site after all, at least in Firefox), I meant that we can give them a little bit more specific/actionable message since `This content can't be viewed embedded in another website.` sounds more like a global issue with the Kibana itself (= aka Kibana is configured in such a way that its content cannot be viewed when embedded). Or is it just me?

> The best way to prevent this from happening is to warn users at the point of embedding a visualisation, which requires authentication, that this will cause issues with privacy settings in modern browsers (Chrome will add this feature too) and that they should only embed visualisation on external websites using anonymous access.

The problem here is that anonymous access also relies on cookies :slightly_smiling_face: There are basically three choices:

* Either user marks Kibana as an exception allowing browser to store it's cookies in the "embedded context" (ideal option for us)
* Or administrator puts an authenticating proxy in front of Kibana (the worst for us as it disables a bunch of features that rely on sessions/cookies).
* Or administrator disables security completely

Reply:
Good point! These are all terrible options tbh. :/ 

I don't think solving this should be in scope of this issue though. Let's take that discussion offline.

Reply:
> This content can't be viewed embedded in another web site.

What does "this content" refer to?

If there is an action for the user to take, then it should be added to the message.

The second message, "Cookies are required..." is nice because it tells the user how to fix the problem.

Reply:
> What does "this content" refer to?
It could be a dashboard or a visualisation but we don't know at this point. (@azasypkin Can anything else be embedded?)

> If there is an action for the user to take, then it should be added to the message.
@azasypkin What action do you think the user can take other than opening the content in a new window?

Reply:
> @azasypkin Can anything else be embedded?

We only provide UI controls to embed visualizations and dashboards, but we know that some users embed other parts of Kibana as well (briefly mentioned [here](https://www.elastic.co/guide/en/kibana/current/kibana-authentication.html#embedded-content-authentication))

> @azasypkin What action do you think the user can take other than opening the content in a new window?

I'd say "open content in a new window or adjust your browser settings to allow 3rd-party cookies", something similar to what Microsoft Teams does. This way we give them an immediate workaround and the direction where to dig to if they want to eventually improve UX. We don't use cookies for ads or something like that, it's an important security tool we need to protect users data, so I feel like it's acceptable. What do you think?


![teams_und_3rd_party_cookies01](https://user-images.githubusercontent.com/1713708/146149517-dfa9ea15-f36a-4151-bfb7-379e051829ab.gif)

Reply:
I like this approach. How about something along these lines? 

<img width="823" alt="Screenshot 2021-12-15 at 11 37 12" src="https://user-images.githubusercontent.com/190132/146179771-9d396c7e-10e5-40ed-afc3-578e7e9f9451.png">

Reply:
Disclaimer: I haven't reviewed the entire PR, just skimming the conversations here, so maybe this has already been answered somewhere else 😄 

One reason that the user could run into this scenario is that they're using an embedded Kibana frame with browser settings that don't allow cross-origin cookies (e.g., Safari out of the box).
A _different_ reason the user could run into this scenario is that Kibana is not setting `SameSite=None` on its cookies.

Can we differentiate these two situations on the client side?
* The former would show this message as written
* The latter would show a slightly different message that says "your administrator needs to allow cross-origin cookies" or something along those lines

We'd need to expose the `xpack.security.sameSiteCookies` setting to the client to differentiate, but that should be doable. WDYT?

Reply:
> I like this approach. How about something along these lines?

Yep, I think that the best we can do here :+1: Regarding `this visualisation` I don't have a good suggestion since we don't really know what we're showing, so anything less specific would work I guess.

> We'd need to expose the xpack.security.sameSiteCookies setting to the client to differentiate, but that should be doable. WDYT?

I'll defer the assessment to Thom, but exposing this setting should be pretty straightforward (supported by the Core), we do this for other config entries already: https://github.com/elastic/kibana/blob/4385ac4d83347be95b6b817611bc2633e44d7965/x-pack/plugins/security/server/index.ts#L41-L44

Reply:
We can't distinguish between the two. In Safari for instance SameSite=None has no effect when prevent cross site tracking is enabled. So in both cases cookies would be disabled but we wouldn't know whether it was a result of prevent cross site tracking or because or same site cookies.

Reply:
> We can't distinguish between the two. In Safari for instance SameSite=None has no effect when prevent cross site tracking is enabled.

Well, we can still change the message appropriately when SameSite=None. But yes, I realize that doesn't help our Safari users. I just meant that we shouldn't always say "it's a browser problem" when it sometimes is because the Kibana server is sending a cookie that should not be accessible in a cross-origin context.

So I just think we need two different messages, that's all.

Reply:
> if the cookie does not have SameSite=None, it will _never_ work in an embedded page (cross-origin context)

Not sure this is true. In Safari I can view embedded content without SameSite=None, only by disabling "Prevent cross-site tracking".

Reply:
> Not sure this is true. In Safari I can view embedded content without SameSite=None, only by disabling "Prevent cross-site tracking".

Good point.

[According to MDN](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Set-Cookie/SameSite#browser_compatibility), Chrome, Edge, and Opera all default to Lax.
And even though that page says Firefox doesn't default to Lax, it looks like that has changed as of a couple weeks ago: https://github.com/mdn/content/issues/10857#issuecomment-982280050

So, all modern, major desktop browsers _except_ Safari behave this way.